karelics_vesc_can_driver/battery_status.py

```diff
@@ -0,0 +1,137 @@
+        return list(filter(None, topic.split(sep='/')))[0]
+
+    def get_vescs_from_status_topics(self, topics_list):
+        active_vescs = []
+
+        for topic in topics_list:
+            active_vescs.append(self.vesc_from_topic(topic))
+
+        return active_vescs
+
+    def create_new_status_sub(self, topic, vesc):
+        status_sub = self.create_subscription(VescStatus, topic, partial(self.vesc_status_cb,
+                                                                         vesc=vesc),
+                                              qos_profile=1)
+        self.vesc_status_subscribers[vesc] = status_sub
+
+    def create_new_vesc_status_subs(self, topics_list):
+        for topic in topics_list:
+            vesc_from_topic = self.vesc_from_topic(topic)
+            if vesc_from_topic not in self.vesc_status_subscribers.keys():
+                self.create_new_status_sub(topic, vesc_from_topic)
+
+    def destroy_stale_vesc_status_subs(self, active_vescs):
+        vescs_to_destroy = []
+        for vesc in list(self.vesc_status_subscribers.keys()):
+            if vesc not in active_vescs:
+                vescs_to_destroy.append(vesc)
+        for vesc in vescs_to_destroy:
+            self.destroy_subscription(self.vesc_status_subscribers[vesc])
+            del self.vesc_status_subscribers[vesc]
+            if self.vesc_voltages.get(vesc):
+                del self.vesc_voltages[vesc]
+
+    def publish_battery_percentage(self):
+        active_vesc_status_topics = self.get_vesc_status_topics()
+        active_vescs = self.get_vescs_from_status_topics(active_vesc_status_topics)
+
+        self.create_new_vesc_status_subs(active_vesc_status_topics)
+
+        self.destroy_stale_vesc_status_subs(active_vescs)
+
+        if len(active_vescs) == 0:
+            return
+
+        mean_battery_voltage = self.get_mean_battery_voltage()
+
+        battery_state = BatteryState()
+        battery_state.voltage = mean_battery_voltage
+        battery_state.percentage = self.get_battery_percentage(mean_battery_voltage)
+        self.battery_pub.publish(battery_state)
+
+    def get_battery_percentage(self, curr_voltage):
+        last_index = len(self.key_voltages)
+        for i in range(last_index):
+            upper_volt = self.key_voltages[i][0]
+            upper_percentage = self.key_voltages[i][1]
+
+            if i == last_index - 1:
+                return float(upper_percentage)
+
+            lower_volt = self.key_voltages[i + 1][0]
+            lower_percentage = self.key_voltages[i + 1][1]
+            if i == 0 and curr_voltage > upper_volt:
+                return float(upper_percentage)
+            elif upper_volt > curr_voltage > lower_volt:
+                percentage = float((curr_voltage - lower_volt) / (upper_volt - lower_volt))
+                scaled_percentage = float(lower_percentage + (upper_percentage - lower_percentage) * percentage)
+                return scaled_percentage
```

Comment:
I would add the float only at return statements, python is generally smart enough to figure out when to use floats